indexer/src/main/scala/scalding/BaseAlternateNamesJoinIntermediateJob.scala

```diff
@@ -0,0 +1,326 @@
+            None
+          }
+        })
+
+        candidates.sortBy(_.length).headOption.toSeq
+      }
+
+      def doRewrites(names: Seq[String]): Seq[String] = {
+        val nameSet = new scala.collection.mutable.HashSet[String]()
+        nameRewritesMap.foreach({case(from, toList) => {
+          names.foreach(name => {
+            toList.foreach(to => {
+              nameSet += from.replaceAllIn(name, to)
+            })
+          })
+        }})
+        nameSet ++= names.map(_.replace("ß", "ss"))
+        nameSet.toSeq
+      }
+
+      lazy val bigDeleteRe = {
+        val re = nameDeletesList
+          .map(_ + "\\b")
+          .sortBy(_.length * -1)
+          .mkString("|")
+        ("(?i)%s".format(re)).r
+      }
+
+      def doDelete(name: String): Option[String] = {
+        val newName = bigDeleteRe.replaceAllIn(name, "")
+        if (newName != name) {
+          Some(fixName(newName))
+        } else {
+          None
+        }
+      }
+
+      def hackName(): Seq[String] = {
+        // HACK(blackmad): TODO(blackmad): move these to data files
+        if (woeType == YahooWoeType.ADMIN1 && cc == "JP" && (lang == "en" || lang == "")) {
+          Seq(name + " Prefecture")
+        } else if (woeType == YahooWoeType.TOWN && cc == "TW" && (lang == "en" || lang == "")) {
+          Seq(name + " County")
+          // Region Lima -> Lima Region
+        } else if (woeType == YahooWoeType.ADMIN1 && cc == "PE" && name.startsWith("Region")) {
+          Seq(name.replace("Region", "").trim + " Region")
+        } else {
+          Nil
+        }
+      }
+
+      def rewriteNames(names: Seq[String]): (Seq[String], Seq[String]) = {
+        val deleteModifiedNames: Seq[String] = names.flatMap(doDelete)
+
+        val deaccentedNames = names.map(NameNormalizer.deaccent).filterNot(n =>
+          names.contains(n))
+
+        val rewrittenNames = doRewrites(names ++ deleteModifiedNames).filterNot(n =>
+          names.contains(n))
+
+        (deaccentedNames, (deleteModifiedNames ++ rewrittenNames).distinct)
+      }
+
+      def isLocalLang() =
+        countryLangMap.getOrElse(cc, Nil).contains(lang)
+
+      def buildFeatureName(name: String, flags: Seq[FeatureNameFlags]) = {
+        FeatureName.newBuilder
+          .lang(lang)
+          .name(name)
+          .flags(flags)
+          .result
+      }
+
+      def processNameList(names: Seq[String], originalFlags: Seq[FeatureNameFlags]): Seq[FeatureName] = {
+        names.map(n => {
+          val finalFlags = originalFlags ++
+            (if (isLocalLang()) {
+              Seq(FeatureNameFlags.LOCAL_LANG)
+            } else {
+              Nil
+            })
+          buildFeatureName(n, finalFlags)
+        })
+      }
+
+      val originalNames = Seq(name)
+      val hackedNames = hackName()
+      val (deaccentedNames, allModifiedNames) = rewriteNames(originalNames)
+      val shortenedNames = doShorten()
+
+      processNameList(originalNames, flags) ++
```

Comment:
this method is too long and has too many closures with nonobvious scope, maybe make it a class